File: flattenBinaryTree.java
public class flattenBinaryTree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public TreeNode recurse(TreeNode node) {
        if (node == null) return node;
        if (node.left == null && node.right == null) return node;

        TreeNode leftTail = recurse(node.left);
        TreeNode rightTail = recurse(node.right);

        if (leftTail != null) {
            leftTail.right = node.right;
            node.right = node.left;
            node.left = null;
        }

        return rightTail != null? rightTail: leftTail;
    }
    
    public void flatten(TreeNode root) {
        recurse(root);
    }
}

File: flattenBinaryTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def recurse(self, node: TreeNode) -> TreeNode:
        if not node:
            return node
        if not node.left and not node.right:
            return node
        
        leftTail = self.recurse(node.left)
        rightTail = self.recurse(node.right)

        if leftTail:
            leftTail.right = node.right
            node.right = node.left
            node.left = None

        return rightTail if rightTail else leftTail


    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return root
        self.recurse(root)

